sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisWatermark.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kinesis;
+
+import java.util.function.BooleanSupplier;
+import org.apache.beam.sdk.transforms.Min;
+import org.apache.beam.sdk.util.MovingFunction;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** Keeps track of current watermark using {@link MovingFunction}. */
+class KinesisWatermark {
+  /** Period of updates to determine watermark. */
+  private static final Duration SAMPLE_UPDATE = Duration.standardSeconds(5);
+
+  /** Period of samples to determine watermark. */
+  static final Duration SAMPLE_PERIOD = Duration.standardMinutes(1);
+
+  /**
+   * Period after which watermark should be updated regardless of number of samples. It has to be
+   * longer than {@link KinesisWatermark#SAMPLE_PERIOD}, so that for most of the cases value
+   * returned from {@link MovingFunction#isSignificant()} is sufficient to decide about watermark
+   * update.
+   */
+  static final Duration UPDATE_THRESHOLD = SAMPLE_PERIOD.multipliedBy(2);
+
+  /** Constant representing the maximum Kinesis stream retention period. */
+  static final Duration MAX_KINESIS_STREAM_RETENTION_PERIOD = Duration.standardDays(7);
+
+  /** Minimum number of unread messages required before considering updating watermark. */
+  static final int MIN_MESSAGES = 10;
+
+  /**
+   * Minimum number of SAMPLE_UPDATE periods over which unread messages should be spread before
+   * considering updating watermark.
+   */
+  private static final int MIN_SPREAD = 2;
+
+  private Instant lastWatermark = Instant.now().minus(MAX_KINESIS_STREAM_RETENTION_PERIOD);
+  private Instant lastUpdate = new Instant(0L);
+  private final MovingFunction minReadTimestampMsSinceEpoch =
+      new MovingFunction(
+          SAMPLE_PERIOD.getMillis(),
+          SAMPLE_UPDATE.getMillis(),
+          MIN_SPREAD,
+          MIN_MESSAGES,
+          Min.ofLongs());
+
+  public Instant getCurrent(BooleanSupplier shardsUpToDate) {
+    Instant now = Instant.now();
+    Instant readMin = getMinReadTimestamp(now);
+    if (readMin == null) {
+      if (shardsUpToDate.getAsBoolean()) {
+        updateLastWatermark(now, now);
```

Comment:
Should this be `now.minus(SAMPLE_PERIOD)` or `now.minus(UPDATE_PERIOD)`? I am guessing 'SAMPLE_PERIOD' is chosen since the message timestamps could vary within that period and not strictly ordered. So if 'shartsUpToDate' is true now, it is expected that next message could have a timestamp between `now` and `now.minus(SAMPLE_PERIOD)`. 

Otherwise, completely idle source will have its watermark set to now, and a very low traffic would have its watermark 2 minutes in the past. May not be major issue.

Reply:
I think that it makes sense, shard iterators determine if they are up to date based on the last call to get records from Kinesis, so it might happen that new records will be behind now causing them to be classified as late records. Following the suggestion from documentation "this value should be as late as possible." using `now.minus(SAMPLE_PERIOD)` looks like good choice here.
@pawel-kaczmarczyk do you have any thoughts on that?

I've run our test pipelines with suggested change to check if it will any negative effect (I suspect it won't). If everything will look good tomorrow I will update the pull request.

Reply:
@rangadi I've updated code to include this change. As suspected, we have not seen any issues with our test pipeline.
Additionally I've changed tests a bit, one of the tests was randomly failing due to incorrect initialisation order.

Reply:
Thanks for the fix and updates. Looks great!